rioxarray/xarray_plugin.py

```diff
@@ -0,0 +1,43 @@
+import os.path
+
+import xarray as xr
+
+from . import _io
+
+CAN_OPEN_EXT = {"geotif", "geotiff", "j2k", "jp2", "tif", "tiff", "vrt"}
+
+
+class GdalBackend(xr.backends.common.BackendEntrypoint):
+    def open_dataset(
+        self,
+        filename_or_obj,
+        drop_variables=None,
+        mask_and_scale=False,
+        parse_coordinates=None,
+        lock=None,
+        masked=False,
+        variable=None,
+        group=None,
+        default_name="band_data",
+    ):
+        ds = _io.open_rasterio(
+            filename_or_obj,
+            mask_and_scale=mask_and_scale,
+            parse_coordinates=parse_coordinates,
+            lock=lock,
+            masked=masked,
+            variable=variable,
+            group=group,
+            default_name=default_name,
+        )
+        if isinstance(ds, xr.DataArray):
+            ds = ds.to_dataset()
+        ds = ds.reset_coords("spatial_ref")
```

Comment:
This should be removed. `spatial_ref` should be a coordinate.

Reply:
I'm trying to follow the CF Conventions where `grid_mapping` refers to a non-coordinate variable:

https://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#grid-mappings-and-projections

and the `coordinates` attribute is intended to be used only for variable containing coordinate values, not the CRS.

I know it is annoying when you work with `DataArray`s that the `spatial_ref` variable is left in the `Dataset`, but that is the specification that most CF-aware tools expect.

By removing the `reset_coords` and saving with `to_netcdf` I obtain a netCDF with:

```
$ ncdump -h test.nc
netcdf test {
dimensions:
...
	short band_data(band, y, x) ;
		band_data:_FillValue = -9999s ;
		band_data:scale_factor = 1. ;
		band_data:add_offset = 0. ;
		band_data:grid_mapping = "spatial_ref" ;
		band_data:coordinates = "spatial_ref" ;
```

Where the `band_data:coordinates = "spatial_ref" ;` looks weird.

What is the reason to keep it in the `.coords`?

Reply:
Thanks for being thorough in your investigation :+1:

> What is the reason to keep it in the .coords?

The main reason is to preserve the CRS information in as non-lossy mechanism as possible that works with GDAL when you do `to_netcdf`. Without it, you cannot store the CRS on a standalone `DataArray` (the default return type with `open_rasterio`).

> that is the specification that most CF-aware tools expect

It works as a coordinate variable with GDAL (the main netCDF CF reading tool rioxarray supports) and ArcMap. Which CF tools does this not work with?

> I'm trying to follow the CF Conventions where grid_mapping refers to a non-coordinate variable

Might be worth raising an issue here for clarification: https://github.com/cf-convention/cf-conventions

Reply:
GDAL reads the CRS information in the `grid_mapping` variable just fine even if it is not a coordinate, I'm using this feature a lot and I just double checked that in fact works with the current implementation as well.

```
>>> ds = xr.open_dataset("test/test_data/input/cog.tif")
>>> ds
<xarray.Dataset>
Dimensions:      (band: 1, x: 500, y: 500)
Coordinates:
  * band         (band) int64 1
  * y            (y) float64 2.715e+06 2.714e+06 ... 2.565e+06 2.565e+06
  * x            (x) float64 1.635e+06 1.635e+06 ... 1.784e+06 1.784e+06
Data variables:
    spatial_ref  int64 ...
    band_data    (band, y, x) int16 ..
>>> ds.to_netcdf("test.nc")
>>> rds = rasterio.open("test.nc")
>>> rds.crs
CRS.from_wkt('PROJCS["Albers",GEOGCS["WGS 84",DATUM["WGS_1984",SPHEROID["WGS 84",6378140,298.256999999996,AUTHORITY["EPSG","7030"]],AUTHORITY["EPSG","6326"]],PRIMEM["Greenwich",0],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AUTHORITY["EPSG","4326"]],PROJECTION["Albers_Conic_Equal_Area"],PARAMETER["latitude_of_center",23],PARAMETER["longitude_of_center",-96],PARAMETER["standard_parallel_1",29.5],PARAMETER["standard_parallel_2",45.5],PARAMETER["false_easting",0],PARAMETER["false_northing",0],UNIT["metre",1,AUTHORITY["EPSG","9001"]],AXIS["Easting",EAST],AXIS["Northing",NORTH]]')
>>> rds.transform
Affine(300.0, 0.0, 1634415.0,
       0.0, -300.0, 2714805.0)
```

Reply:
> Which CF tools does this not work with?

I can't reproduce an error ATM, but CDO is usually quite picky about the content of the `coordinates` attributes.

Reply:
A big issue for not including the CRS in the coordinates is the lack of support for the `DataArray`. The CRS information will be lost for any `data_var` selected from a `Dataset` and cannot be set when creating a `DataArray` from scratch. Without the CRS information, several features in the `rio` extension won't work properly that require the CRS on the `DataArray` such as `rio.to_raster`, `rio.clip`, and `rio.reproject` to name a few.

Reply:
This is morphing into a discussion on the *xarray* data model and on the limitations of `xr.DataArray`. At them moment coordinate variables are special cased and associated with the `DataArray` while other auxiliary variables, like `grid_mapping` and `ancillary_variables ` variables `, are not.

I understand the current need for the `spatial_ref`-variable-as-coordinate hack, but I'm concerned about the fact that the resulting in-memory and on-disk data structure do not conform to the well known standard.

We should probably move the conversation on pydata/xarray.

Reply:
OK, I had completely missed https://github.com/pydata/xarray/pull/2844 and I'm reading the other issues `grid_mapping` as well, then I'll reassess.

Reply:
So technically adding support for `decode_coords="all"` in the plugin and setting `.encoding` accordingly should make everybody happy.

Note that the default would be to open with `decode_coords=True` equals to `decode_coords="coordinates"` that is similar to applying `.reset_coords`, but you can have an improved `spatial_ref`-variable-as-coordinate in-memory that also saves a correct netCDF on-disk. Cool.

Reply:
@alexamici, I appreciate your thoughts and ideas on the on the matter. For now, storing the Coordinate Reference System in the coordinate versus the data variables is something I would like to discuss in a separate issue and implement any changes in a separate PR for clarity in the changeset and to allow for more thought and discussion on the subject. Would you mind removing the logic from this PR and create a separate issue?